plutus-core/untyped-plutus-core/Language/UntypedPlutusCore/Evaluation/HOAS.hs

```diff
@@ -0,0 +1,307 @@
+-- is annotated with. We could probably handle annotations better, but we don't care for now.
+fromHTerm (HLamAbs ann name body) = LamAbs ann name <$> (body (HVar ann name) >>= fromHTerm)
+fromHTerm (HApply ann fun arg)    = Apply ann <$> fromHTerm fun <*> fromHTerm arg
+fromHTerm (HDelay ann getBody)    = Delay ann <$> (getBody () >>= fromHTerm)
+fromHTerm (HForce ann term)       = Force ann <$> fromHTerm term
+fromHTerm (HError ann)            = pure $ Error ann
+
+-- | Convert a 'Value' into a 'Term' running all internal monadic actions along the way and
+-- extracting all partial builtin applications.
+fromValue
+    :: Value unique name uni fun ann
+    -> EvalM unique name uni fun ann (Term name uni fun ann)
+fromValue = fromHTerm >=> bindFunM (const _builtinAppTerm)
+
+-- | Run an 'EvalM' computation and convert the cause of a possible error from 'Value' to 'Term'.
+runEvalM
+    :: EvalM unique name uni fun ann a
+    -> Either (HoasException fun (Term name uni fun ann)) a
+runEvalM = bimap errorValueToTerm id . unEvalM where
+    -- Here we call 'runEvalM' recursively. It's fine when the underlying monad is 'Either',
+    -- but if it had 'ReaderT', then we'd also need to make sure that 'runEvalM' is supplied
+    -- with the most recent environment, not the initial one.
+    errorValueToTerm = either id id . traverse (runEvalM . fromValue)
+
+lookupVar
+    :: HasUnique name unique
+    => ann -> name -> UniqueMap unique value -> EvalM unique name uni fun ann value
+lookupVar ann name env =
+    case lookupName name env of
+        Just term -> pure term
+        Nothing   ->
+            throwingWithCause _InternalHoasError FreeVariableHoasError $
+                Just $ HVar ann name
+
+-- | Retrieve the meaning of a built-in function.
+lookupBuiltin
+    :: (value ~ Value unique name uni fun ann, Ix fun)
+    => ann -> fun -> BuiltinsRuntime fun value -> EvalM unique name uni fun ann value
+lookupBuiltin ann fun (BuiltinsRuntime meanings) =
+    case meanings ^? ix fun of
+        Nothing      -> throwingWithCause _InternalHoasError (UnknownBuiltinHoasError fun) Nothing
+        Just meaning -> pure . HBuiltin ann $ BuiltinApp (pure $ Builtin ann fun) meaning
+{- Note [Handling non-constant results]
+Evaluation may return a non-constant term. This has a couple of implications:
+
+1. we have to keep a 'Term' representation of a partial builtin application, so that if evaluation
+   results in an underapplied builtin like @addinteger 1@, we can return that term. If we were only
+   to keep the denotation of a builtin, we couldn't reconstruct the term from it
+2. 'evalApply' and 'evalForce' have to handle the case when their argument is neither a
+   'HLamAbs'/'HDelay' nor a built-in function, because if evaluation results in, say,
+   @\f -> f 1@, then we need to turn that application into an 'HApply' AST node (which itself
+   can be forced or applied to another term), which upon final reification becomes an 'Apply' node.
```

Comment:
Is that not what 2 tries to say? 2 refers to `evalApply` and there's this line:

```haskell
evalApply ann fun                arg = pure $ HApply ann fun arg
```

> which I think means you can end up with stuck terms because you'll get a variable applied to something.

Yeah, but can that only happen during the final reification? In order to encounter a variable that maps to a term with "non-lookupable" variables in it you need to look under a lambda that you keep (as opposed to evaluate), which is what makes the variable that it binds "non-lookuable", -- and that only happens during the final reification. (sorry for this sentence, I did try to make it more readable)

Reply:
I was thinking about `fromHTerm` where you actually do the NBE thing, but actually I think that's not relevant here.

Reply:
I'm now confused. Could you either resolve the comment or explain a bit more what I should do about it?

Reply:
I think it's fine, I was wrong.